### PR TITLE
Refactored provider classes from client.ts to providers.ts

### PR DESCRIPTION
**Pull request by spacemonkd**
Refactored Provider classes ( **FoldingRangeProvider, SemanticTokensProvider, DocumentFormattingEditProvider, DocumentRangeFormattingEditProvider, OnTypeFormattingEditProvider** ) from client.ts to providers.ts.

Pull request for issue: [#5871 ](https://github.com/microsoft/vscode-cpptools/issues/5871)

Kindly review and suggest necessary changes @bobbrow @Colengms @michelleangela

### CODE REVIEW

**Comment by spacemonkd** [comment]
Azure pipelines giving `missing file header (tslint:file-header)  @typescript-eslint/tslint/config` error. Any suggestion as to why this is happening?
@bobbrow @Colengms @michelleangela

**Comment by bobbrow** [comment]
Hi @devabhishekpal,

Thanks for taking a look at this issue.  It was assigned to @Colengms as I think he had some ideas as to how he wanted it structured.  I will defer to him as to whether this is the direction he had in mind.

**Comment by spacemonkd** [comment]
> Hi @devabhishekpal,
> 
> Thanks for taking a look at this issue. It was assigned to @Colengms as I think he had some ideas as to how he wanted it structured. I will defer to him as to whether this is the direction he had in mind.

Thank you dear sir.

**Comment by sean-mcmanus** [comment]
> Azure pipelines giving `missing file header (tslint:file-header) @typescript-eslint/tslint/config` error. Any suggestion as to why this is happening?
> @bobbrow @Colengms @michelleangela

I think that error is due to missing the following at the top of provider.ts
```
/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All Rights Reserved.
 * See 'LICENSE' in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
```.

**Comment by spacemonkd** [comment]
Ok sir making changes and committing.

**Comment by Colengms** [comment]
Hi @devabhishekpal .  Thanks a lot for jumping in and helping us out.  :)   The only thing I would have done differently is create a separate file for each provider.  We'd also like to move DocumentSymbolProvider, WorkspaceSymbolProvider, FindAllReferencesProvider, and RenameProvider as well.  Do you want to tackle these also?  :)   If not, we could use what is here, and deal with the rest later.

**Comment by spacemonkd** [comment]
I am on it @Colengms

**Comment by spacemonkd** [comment]
Could you check up on the variables? @Colengms 
I created classes and exported them making the variables public static so that we can refernce the same value across file.

Not sure if that is desirable so could you kindly check on it.

export class AbortRequestIdHolder{
    abortRequestId;
}

export class ReferenceParamsHolder{
  referencesPendingCancellations;
  referencesParams;
  referencesRequestPending;
}

andd

export class RenameParamsHolder{
   renameRequestsPending;
   renamePending;
}

these were originally declared as const values.

**Comment by spacemonkd** [comment]
Can you please check the commits if possible 😅 @bobbrow @Colengms

**Comment by spacemonkd** [comment]
Can someone check the updated code and give a review or changes needed? @bobbrow @sean-mcmanus @Colengms @Colengms @michelleangela

**Comment by michelleangela** [comment]
@devabhishekpal 
Thank you for the update. Please note that the team may not be able to review the changes until time permits and based on other higher priority issues and tasks that are also currently being worked on.

**Comment by spacemonkd** [comment]
@michelleangela thank you for informing. It is alright. As time permits any review of the code would be really helpful, and thanks once again😊

**Comment by Colengms** [comment]
@devabhishekpal Instead of 'Holder' classes, could you make these individually exported fields for now?  We eventually need to reorganize those vars.  Also, could you move all providers into a new providers subdirectory?

**Comment by spacemonkd** [comment]
Are those used globally? Like normally exporting them would prevent any value modified in other file from reflecting upon it isnt it?😅. @Colengms

**Comment by Colengms** [comment]
> Are those used globally? Like normally exporting them would prevent any value modified in other file from reflecting upon it isnt it?😅. 

@devabhishekpal Are you saying that exporting a variable will result in each file seeing a different version of that variable?  I don't believe that is the case.  You should be able to export a variable.  If there is an issue with this, I'm not aware of it.  Can you confirm?

**Comment by spacemonkd** [comment]
> > Are those used globally? Like normally exporting them would prevent any value modified in other file from reflecting upon it isnt it?😅. 
> 
> 
> 
> @devabhishekpal Are you saying that exporting a variable will result in each file seeing a different version of that variable?  I don't believe that is the case.  You should be able to export a variable.  If there is an issue with this, I'm not aware of it.  Can you confirm?

Sorry, my fault. Imported values do reflect upon across files. 😅

Will make the necessary changes. And create separate files😊 @Colengms

**Comment by spacemonkd** [comment]
Changes done @Colengms 

**Note:** Used two imports as renameParamsHolder and referencesParamsHolder, where they are referenced to avoid confusion and increase readability. They might be redundant and if told will merge them into a single Holder variable.

**Comment by spacemonkd** [comment]
Linter shows that the variables must be const as they arent reassigned, but I am changing the values in the provider files. Like incrementing or reassigning new values. What shall I do? @Colengms @bobbrow 

Is it reassignment it I am using a reference to the variable like 

**renameParamsHolder.renameReques = \<value\>** ?

Because const isnt reassignable but it is mutable isn't it?

**Comment by sean-mcmanus** [comment]
> Linter shows that the variables must be const as they arent reassigned, but I am changing the values in the provider files. Like incrementing or reassigning new values. What shall I do? @Colengms @bobbrow
> 
> Is it reassignment it I am using a reference to the variable like
> 
> **renameParamsHolder.renameReques = <value>** ?
> 
> Because const isnt reassignable but it is mutable isn't it?

Yeah, const should be used when a TypeScript variable is not reassigned, but it can still be changed. So renameParamsHolder would be const, but renameReuest would not be.

**Comment by sean-mcmanus** [comment]
> Yeah, const should be used when a TypeScript variable is not reassigned, but it can still be changed. So renameParamsHolder would be const, but renameReuest would not be.

Oh, but it looks like the linter is complaining about the variable that is being re-assigned. Do you think there's a linter bug?

**Comment by sean-mcmanus** [comment]
> Oh, but it looks like the linter is complaining about the variable that is being re-assigned. Do you think there's a linter bug?

Oh, the linter is complaining about variables like `export let referencesRequestPending: boolean = false;` so those should be changed to const.

**Comment by sean-mcmanus** [comment]
> Oh, the linter is complaining about variables like `export let referencesRequestPending: boolean = false;` so those should be changed to const.

Oh, but that is the same variable that is being reassigned. Yeah, looks like the linter shouldn't be showing that const error for exported values that are reassigned. Maybe use // eslint-disable-line as a workaround.

**Comment by sean-mcmanus** [comment]
> Oh, but that is the same variable that is being reassigned. Yeah, looks like the linter shouldn't be showing that const error for exported values that are reassigned. Maybe use // eslint-disable-line as a workaround.

I filed an eslint bug report at https://github.com/eslint/eslint/issues/13621

**Comment by sean-mcmanus** [comment]
> I filed an eslint bug report at [eslint/eslint#13621](https://github.com/eslint/eslint/issues/13621)

Adding 
// eslint-disable-next-line prefer-const
or
// eslint-disable-line prefer-const
should be a workaround.

**Comment by spacemonkd** [comment]
Ok @sean-mcmanus sir. Thank you😊

**Comment by spacemonkd** [comment]
Raised issue as at eslint they told to forward issue to[@typescript-eslint/parser](https://github.com/typescript-eslint/typescript-eslint/issues/2426) @sean-mcmanus

**Comment by Colengms** [comment]
@devabhishekpal Would it work to move all of these `export let` variables to public static member variables in DefaultClient?

**Comment by sean-mcmanus** [comment]
> @devabhishekpal Would it work to move all of these `export let` variables to public static member variables in DefaultClient?

You might also look into using get/set methods -- not sure if those work on static members though. If so, then the syntax would be like a variable access.

**Comment by spacemonkd** [comment]
> Please ensure you read the instructions in the template before filing an issue:
> 
> ```
> 
> ```
> 
> The `prefer-const` rule cannot detect usages across files. ESLint works on single files only.
> Even with type information we couldn't do that.
> 
> Exporting a `let` is an anti-pattern as well. We cannot detect how you are using the module, so there's no way for us to write a lint rule which allows you to do this anti-pattern.
> 
> See my answer in your other issue: [eslint/eslint#13621 (comment)](https://github.com/eslint/eslint/issues/13621#issuecomment-681188919)

This is what they said for typescript-eslint

**Comment by spacemonkd** [comment]
> @devabhishekpal Would it work to move all of these `export let` variables to public static member variables in DefaultClient?

Okk... I am shifting those to DefaultClient class. Shall I also move **const ReferencesPendingCancellations**? Then we can refer them in a single object.

**Comment by spacemonkd** [comment]
> > @devabhishekpal Would it work to move all of these `export let` variables to public static member variables in DefaultClient?
> 
> You might also look into using get/set methods -- not sure if those work on static members though. If so, then the syntax would be like a variable access.

I guess we can use getters and setters with static variables. I will try that

**Comment by devabhishekpal** [comment]
There are some problems with using getter and setters @sean-mcmanus sir. 

I have used two three files to emulate the work in our case.

**importer.ts**

```
export class Importer {
  public _a: number = 0;

  public get a() {
    return this._a;
  }

  public set a(aVal: number) {
    this._a = aVal;
  }
}
```
**exporter.ts**

```
import { Importer } from "./importer";

export class Exporter {
  public importer = new Importer();

  public viewA() {
    console.log(this.importer.a);
  }

  public IncrementA() {
    this.importer.a;
  }
}
```

**main.ts**

```
import { Exporter } from "./exporter";
import { Importer } from "./importer";

let exporter = new Exporter();
let importer = new Importer();

exporter.viewA();
exporter.IncrementA();

exporter.viewA();
console.log(importer.a);
```

Now this is causing the value of a to increment with respect to exporter.ts but the values arent carrying over to **a** in importer.ts

output

Look at the output.

**exporter.viewA()** is giving **0**.
**exporter.IncrementA()** is incrementing the value to **1**.

But when I am calling **importer.a** the value in there is still **0**.

**Comment by spacemonkd** [comment]
There are some problems with using getter and setters @sean-mcmanus sir, @Colengms  sir, @bobbrow sir. 

I have used two three files to emulate the work in our case.

**importer.ts**

```
export class Importer {
  public _a: number = 0;

  public get a() {
    return this._a;
  }

  public set a(aVal: number) {
    this._a = aVal;
  }
}
```
**exporter.ts**

```
import { Importer } from "./importer";

export class Exporter {
  public importer = new Importer();

  public viewA() {
    console.log(this.importer.a);
  }

  public IncrementA() {
    this.importer.a;
  }
}
```

**main.ts**

```
import { Exporter } from "./exporter";
import { Importer } from "./importer";

let exporter = new Exporter();
let importer = new Importer();

exporter.viewA();
exporter.IncrementA();

exporter.viewA();
console.log(importer.a);
```

Now this is causing the value of a to increment with respect to exporter.ts but the values arent carrying over to **a** in importer.ts

output

Look at the output.

**exporter.viewA()** is giving **0**.
**exporter.IncrementA()** is incrementing the value to **1**.

But when I am calling **importer.a** the value in there is still **0**.

**Comment by spacemonkd** [comment]
I thought of using static variables but then we cant use getters or setters because without an object we cant call them and static variables aren't based on objects.

**Comment by sean-mcmanus** [comment]
> I thought of using static variables but then we cant use getters or setters because without an object we cant call them and static variables aren't based on objects.

Does following Colen's suggestion of using "public static member variables in DefaultClient" work?

**Comment by spacemonkd** [comment]
I guess that will work. Then we can refer to them with DefaultClient

**Comment by spacemonkd** [comment]
Sorry. Was having connectivity issues for a few days. Will commit within the day🙂

**Comment by spacemonkd** [comment]
After pulling some extra files were also added idk why. But I changed the params to static.

**Comment by sean-mcmanus** [comment]
> After pulling some extra files were also added idk why. But I changed the params to static.

Your last commit changed all the line endings from LF to CLRF. Might be related to https://stackoverflow.com/questions/1967370/git-replacing-lf-with-crlf . Can you undo that commit and don't check in the line ending changes?

**Comment by spacemonkd** [comment]
Okkk

**Comment by sean-mcmanus** [comment]
> Okkk

The conflict is caused by change https://github.com/microsoft/vscode-cpptools/pull/6068/files -- the DocumentFormatRequest got changed to 3 messages. The binary changes to handle the new messages hasn't been published yet, but it should be available next week in 0.30.0-insiders5.

**Comment by spacemonkd** [comment]
Is the current code alright or do I need to make modifications?

**Comment by sean-mcmanus** [comment]
> Is the current code alright or do I need to make modifications?

Some modifications are needed to resolve the conflicts. References to DocumentFormatRequest need to changed to FormatDocumentRequest, FormatRangeRequest, or FormatOnTypeRequest, depending on the context, and
`const DocumentFormatRequest: RequestType<FormatParams, TextEdit[], void, void> = new RequestType<FormatParams, TextEdit[], void, void>('cpptools/format');` should be changed to
```
const FormatDocumentRequest: RequestType<FormatParams, TextEdit[], void, void> = new RequestType<FormatParams, TextEdit[], void, void>('cpptools/formatDocument');
const FormatRangeRequest: RequestType<FormatParams, TextEdit[], void, void> = new RequestType<FormatParams, TextEdit[], void, void>('cpptools/formatRange');
const FormatOnTypeRequest: RequestType<FormatParams, TextEdit[], void, void> = new RequestType<FormatParams, TextEdit[], void, void>('cpptools/formatOnType');
```

File: Extension/src/LanguageServer/client.ts
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/client.ts]: Why did spacing change from 4 spaces to 2?

File: Extension/src/LanguageServer/client.ts
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/client.ts]: Why are some template args spread over multile lines?

**Review by sean-mcmanus — CHANGES_REQUESTED** [review]

**Comment by spacemonkd** [comment]
Okay

**Comment by spacemonkd** [comment]
Required changes are done and the line spacing was an error on my end with auto formatting. Sorry about that😅

**Comment by sean-mcmanus** [comment]
@devabhishekpal There's a build error on line 1360 of client.ts -- looks like DefaultClient.renamePending should be used. I resolved a merge conflict...not sure if that caused it.

**Comment by spacemonkd** [comment]
Okayyy I will check it

**Comment by sean-mcmanus** [comment]
There are some linter issues: 
/home/vsts/work/1/s/Extension/src/LanguageServer/client.ts
  2618:1   error  Expected indentation of 16 spaces but found 14  @typescript-eslint/indent
  2619:1   error  Expected indentation of 16 spaces but found 14  @typescript-eslint/indent
  2619:33  error  Unexpected trailing comma                       comma-dangle

**Comment by spacemonkd** [comment]
Yeah I undated those just now @sean-mcmanus sir

File: Extension/src/LanguageServer/Providers/documentFormattingEditProvider.ts
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/Providers/documentFormattingEditProvider.ts]: There's a double newline here and in a few other files. If you feel like it, you could change the existing "no-multiple-empty-lines" in .eslintrc.js to be "no-multiple-empty-lines": ["error", { "max": 1, "maxEOF": 1, "maxBOF": 0}].
If not, we could do that in a separate PR.

**Review by sean-mcmanus — APPROVED** [review]

**Comment by sean-mcmanus** [comment]
FYI, we're going to delay checking this in until we can merge to release for 0.30.0-insiders5 and 1.0.0...the merge might be in a few hours though...or maybe tomorrow.

File: Extension/src/LanguageServer/Providers/documentFormattingEditProvider.ts
Comment (spacemonkd) [reply Extension/src/LanguageServer/Providers/documentFormattingEditProvider.ts]: Okay sir

**Comment by spacemonkd** [comment]
Okayy

**Comment by spacemonkd** [comment]
I changed the eslint rules. I will fix the multiple spaces on other files by evening

**Comment by spacemonkd** [comment]
Sorry about so many commits...I was editing from my phone so I had to edit each file individually... I have fixed the errors....Kindly check the same.. @sean-mcmanus sir

**Comment by sean-mcmanus** [comment]
> Sorry about so many commits...I was editing from my phone so I had to edit each file individually... I have fixed the errors....Kindly check the same.. @sean-mcmanus sir

Cool, thanks. We might be able to check this in tomorrow after we merge for 1.0 (not sure if @Colengms had any more review feedback).

**Comment by spacemonkd** [comment]
Okayy

**Comment by spacemonkd** [comment]
Is there any other issue for me to look at? @sean-mcmanus sir

**Comment by sean-mcmanus** [comment]
> Is there any other issue for me to look at? @sean-mcmanus sir

No issues. We can probably check it in Monday. We've been busy preparing our 1.0 release.

**Comment by spacemonkd** [comment]
Okayy sir

**Review by Colengms — APPROVED** [review]

**Review by sean-mcmanus — APPROVED** [review]

**Pull request merged**
